File: packages/paygate-python/paygate/utils/nonce.py
"""Nonce + short token helpers. All nonces are bound to a requirements
digest so a client cannot replay a nonce against a different requirement.
"""

from __future__ import annotations

import hashlib
import secrets
import time

try:
    import ulid  # type: ignore[import-not-found]
except ImportError:  # pragma: no cover — ulid-py is a runtime dependency.
    ulid = None  # type: ignore[assignment]


def generate_nonce(digest: str) -> str:
    """Generate a server-issued nonce bound to the requirements digest.

    The ULID prefix gives us sortable, monotonic IDs; the hex suffix binds
    the nonce to the exact digest we signed, so a client cannot swap the
    requirement they're paying for.
    """
    if ulid is None:  # pragma: no cover
        raise RuntimeError(
            "ulid-py is required for generate_nonce; install `paygate[all]`"
        )
    nonce_id = ulid.new().str  # type: ignore[attr-defined]
    random_bytes = secrets.token_hex(8)
    suffix = hashlib.sha256(
        f"{nonce_id}|{digest}|{random_bytes}".encode("utf-8")
    ).hexdigest()[:16]
    return f"{nonce_id}.{suffix}"


def epoch_seconds() -> int:
    """Epoch seconds — safe for cross-process TTL comparisons."""
    return int(time.time())


def short_token() -> str:
    """Short, log-safe token for request correlation."""
    return secrets.token_hex(6)


__all__ = ["epoch_seconds", "generate_nonce", "short_token"]

File: packages/limen-node/src/utils/nonce.ts
import { createHash, randomBytes } from 'node:crypto';
import { ulid } from 'ulid';

/**
 * Generate a server-issued nonce bound to the requirements digest.
 * The ULID gives us sortable, monotonically-increasing ids; the digest
 * suffix binds the nonce to the exact requirements we issued.
 */
export function generateNonce(digest: string): string {
  const id = ulid();
  const rand = randomBytes(8).toString('hex');
  const suffix = createHash('sha256')
    .update(`${id}|${digest}|${rand}`)
    .digest('hex')
    .slice(0, 16);
  return `${id}.${suffix}`;
}

/**
 * Monotonic clock wrapper.  Never use Date.now() for TTL.  This keeps the
 * clock impervious to NTP adjustments within a process lifetime.
 */
export function monotonicSeconds(): number {
  // process.hrtime.bigint() returns nanoseconds since some fixed reference.
  // We map it to seconds, but also need absolute epoch seconds for
  // cross-process comparison (validUntil is epoch seconds).  For that we
  // anchor on the process start time.
  const nsSinceStart = process.hrtime.bigint();
  return Number(nsSinceStart / 1_000_000_000n) + PROCESS_EPOCH_SECONDS;
}

const PROCESS_EPOCH_SECONDS = Math.floor(Date.now() / 1000) - process.uptime();

/** Epoch seconds (safe for cross-process TTL comparisons). */
export function epochSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

/** Short, log-safe token for correlation. */
export function shortToken(): string {
  return randomBytes(6).toString('hex');
}
